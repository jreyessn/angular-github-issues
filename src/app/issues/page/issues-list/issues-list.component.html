<div class="flex flex-col w-full">
    <div>
        <div class="text-3xl font-semibold tracking-tight leading-8">Buscador de Issues</div>
        <div class="font-medium tracking-tight text-secondary">Realice busquedas din√°micas de cualquier Issue en los repositorios publicos de Github</div>
    </div>
    <div class="mt-4">
        <div class="flex gap-4 items-center">
            <app-issue-search-field class="w-full" (search)="search($event)"></app-issue-search-field>
        </div>
        <div>
            <span class="text-sm font-semibold">{{ total$ | async | number }} Issues</span>
        </div>
    </div>
</div>

<div *ngIf="data$ | async as data; else loading" 
     [ngClass]="{
         'grid-cols-1 sm:grid-cols-2 lg:grid-cols-3': data.length > 0,
         'grid-cols-1': data.length == 0
     }"
     class="grid gap-8 w-full mt-4">
     <div *ngIf="data.length == 0" class="flex justify-center w-full">
         <p class="text-secondary">No se han encontrado Issues</p>
     </div>
     <app-issue-card *ngFor="let issueItem of data" [item]="issueItem"></app-issue-card>
</div>

<ng-template #loading>
    <div class="flex justify-center items-center w-full h-72">
        <mat-spinner></mat-spinner>
    </div>
</ng-template>
