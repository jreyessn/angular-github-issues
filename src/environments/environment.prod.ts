export const environment = {
  production: true,
  api_github: "https://api.github.com/"
};
