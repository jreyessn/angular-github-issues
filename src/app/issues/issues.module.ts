import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IssuesListComponent } from './page/issues-list/issues-list.component';
import { RouterModule } from '@angular/router';
import { issuesRoutes } from './issues.routing';
import { IssueCardComponent } from './components/issue-card/issue-card.component';
import { IssueSearchFieldComponent } from './components/issue-search-field/issue-search-field.component';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';
import { MatToolbarModule } from '@angular/material/toolbar';
import { MatIconModule } from '@angular/material/icon';
import { MatButtonModule } from '@angular/material/button';
import { MatInputModule } from '@angular/material/input';
import { MatTooltipModule } from '@angular/material/tooltip';
import { IssuesService } from './services/issues.service';
import { HttpClientModule } from '@angular/common/http';
import { MapRepoPipe } from './pipes/map-repo.pipe';
import {MatProgressSpinnerModule} from '@angular/material/progress-spinner'; 
import {MatSelectModule} from '@angular/material/select';

@NgModule({
  declarations: [
    IssuesListComponent,
    IssueCardComponent,
    IssueSearchFieldComponent,
    MapRepoPipe,
  ],
  imports: [
    // Core Modules
    CommonModule,
    RouterModule.forChild(issuesRoutes),
    FormsModule,
    ReactiveFormsModule,
    
    // Angular Material
    MatToolbarModule,
    MatIconModule,
    MatButtonModule,
    MatInputModule,
    MatTooltipModule,
    MatProgressSpinnerModule,
    MatSelectModule,
    
    // Http
    HttpClientModule
  ],
  providers: [
    IssuesService
  ]
})
export class IssuesModule { }
