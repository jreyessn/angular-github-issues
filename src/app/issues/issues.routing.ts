import { Route } from "@angular/router";
import { IssuesListComponent } from "./page/issues-list/issues-list.component";

export const issuesRoutes: Route[] = [
    {
        path     : '',
        component: IssuesListComponent
    }
];