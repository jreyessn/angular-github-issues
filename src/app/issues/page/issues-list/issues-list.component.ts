import { Component, OnInit } from '@angular/core';
import { BehaviorSubject, map, Observable, of, tap } from 'rxjs';
import { Issue, IssueResponse } from '../../issues.type';
import { IssuesService } from '../../services/issues.service';

@Component({
  selector: 'app-issues-list',
  templateUrl: './issues-list.component.html',
  styleUrls: ['./issues-list.component.scss']
})
export class IssuesListComponent {

  /**
   * Limit records
   */
  limitRecords: number = 12;

  /**
   * Data api
   */
  data$: Observable<Issue[]> = of([])

  /**
   * Total count
   */
  total$: BehaviorSubject<number> = new BehaviorSubject(0)

  constructor(
    private _issueServices: IssuesService
  ) { }

  /**
   * Realiza busqueda
   * @param text TÃ©rmino de busqueda
   */
  search(text: string): void {
    this.data$ = this._issueServices.list(text).pipe(
      tap((res: IssueResponse) => this.total$.next(res.total_count)),
      map((res: IssueResponse) => res.items)
    )
  }

}
