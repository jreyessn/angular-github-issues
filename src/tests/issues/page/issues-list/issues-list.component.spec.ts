import { ComponentFixture, TestBed } from '@angular/core/testing';
import { IssuesService } from '../../../../app/issues/services/issues.service';

import { IssuesListComponent } from '../../../../app/issues/page/issues-list/issues-list.component';
import { HttpClientTestingModule, HttpTestingController } from '@angular/common/http/testing';
import { NO_ERRORS_SCHEMA } from '@angular/core'
import { environment } from '../../../../environments/environment';

jest.setTimeout(15000)

describe('IssuesListComponent', () => {
  let component: IssuesListComponent;
  let fixture: ComponentFixture<IssuesListComponent>;
  let compiled: HTMLElement;
  let service: IssuesService;
  let httpMock: HttpTestingController;

  beforeEach(async () => {
    await TestBed.configureTestingModule({
      imports: [
        HttpClientTestingModule,
      ],
      declarations: [ 
        IssuesListComponent,
      ],
      providers: [ IssuesService ],
      schemas: [ NO_ERRORS_SCHEMA ]
    })
    .compileComponents();
  });

  beforeEach(() => {
    fixture   = TestBed.createComponent(IssuesListComponent);
    component = fixture.componentInstance;
    service   = TestBed.inject(IssuesService)
    httpMock  = TestBed.inject(HttpTestingController)

    fixture.detectChanges();
    compiled = fixture.nativeElement;
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it("Debe de hacer match con el snapshot", () => {
    expect(compiled).toMatchSnapshot()
  })

  it("Debe mostrar el texto de que no se han encontrado Issues", () => {
    const p = compiled.querySelector(".grid p.text-secondary")
    expect(p?.textContent).toContain("No se han encontrado Issues")
  })

  // it("Debe de mostrar la cantidad de Issues traidos", (done) => {
  //   const items = [
  //     { url: "https://github.com/facebook/jest/issues/11607" },
  //     { url: "https://github.com/facebook/jest/issues/11607" },
  //     { url: "https://github.com/facebook/jest/issues/11607" },
  //     { url: "https://github.com/facebook/jest/issues/11607" },
  //     { url: "https://github.com/facebook/jest/issues/11607" },
  //   ]

  //   component.search("jest")
  //   component.data$.subscribe()
  //   const request = httpMock.expectOne(environment.api_github + 'search/issues?q=jest')
  //   request.flush({
  //     total_count:        5000,
  //     incomplete_results: true,
  //     items,
  //   })

  //   fixture.whenStable().then(() => { // wait for your async data
  //     fixture.detectChanges(); // refresh your fake template
  //     compiled = fixture.nativeElement;

  //     const p = compiled.querySelector("#count-issues")
  //     expect(p?.textContent).toContain("5,000 Issues")
      
  //     done()
  
  //   })



  // })


});
