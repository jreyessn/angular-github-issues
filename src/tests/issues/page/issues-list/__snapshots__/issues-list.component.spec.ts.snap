// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IssuesListComponent Debe de hacer match con el snapshot 1`] = `
<div
  id="root1"
>
  <div
    class="flex flex-col w-full"
  >
    <div>
      <div
        class="text-3xl font-semibold tracking-tight leading-8"
      >
        Buscador de Issues
      </div>
      <div
        class="font-medium tracking-tight text-secondary"
      >
        Realice busquedas din√°micas de cualquier Issue en los repositorios publicos de Github
      </div>
    </div>
    <div
      class="mt-4"
    >
      <div
        class="flex gap-4 items-center"
      >
        <app-issue-search-field
          class="w-full"
        />
      </div>
      <div>
        <span
          class="text-sm font-semibold"
          id="count-issues"
        >
          0 Issues
        </span>
      </div>
    </div>
  </div>
  <div
    class="grid gap-8 w-full mt-4 grid-cols-1"
  >
    <div
      class="flex justify-center w-full"
    >
      <p
        class="text-secondary"
      >
        No se han encontrado Issues
      </p>
    </div>
    
    
  </div>
  
  
</div>
`;
